### PR TITLE
Multi env refactor

### PR DESCRIPTION
**Pull request by djbyrne**
# Before submitting

- [X] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [X] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [X] Did you make sure to update the docs?   
- [X] Did you write any new necessary tests?  

## What does this PR do?
This PR is a relatively large refactor, specifically in how the ExperienceSource works. The interface remains roughly the same as before, however, the ExperienceSource now contains a stepper() function which acts as an iterator which continuously yields new experience steps. 

As well as this, the ExperienceSource has been refacfored to allow for multiple environments to be run once. This is an important feature as many algorithms require it and it allows RL models to run with a larger batch size. 

Finally there are some general clean up changes to the models and tests. 

Fixes # 88.

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
:+1:

### CODE REVIEW

File: tests/datamodules/test_experience_sources.py [hunk 1]
```diff
@@ -32,123 +39,285 @@ def test_iterator(self):
...
         self.net = Mock()
         self.agent = DummyAgent(net=self.net)
-        self.env = gym.make("CartPole-v0")
-        self.n_step = 2
-        self.source = NStepExperienceSource(self.env, self.agent, n_steps=self.n_step)
+        self.env = [gym.make("CartPole-v0") for _ in range(2)]
+        self.device = torch.device('cpu')
+        self.source = ExperienceSource(self.env, self.agent)
+
+    def test_init(self):
```
Comment (Borda) [inline tests/datamodules/test_experience_sources.py:68]:
```suggestion
    def test_init_source(self):
```
is it?

File: tests/datamodules/test_experience_sources.py [hunk 2]
```diff
@@ -32,123 +39,285 @@ def test_iterator(self):
...
+        self.assertIsInstance(self.source.pool, list)
+
+        self.assertEqual(len(self.source.states), len(self.source.pool))
+        self.assertEqual(len(self.source.histories), len(self.source.pool))
+        self.assertEqual(len(self.source.cur_rewards), len(self.source.pool))
+        self.assertEqual(len(self.source.cur_steps), len(self.source.pool))
+
+    def test_init_single_env(self):
+        """Test that if a single env is passed that it is wrapped in a list"""
+        single_env = Mock()
```
Comment (Borda) [inline tests/datamodules/test_experience_sources.py:80]: you already have it as `self.net`

File: tests/datamodules/test_experience_sources.py [hunk 3]
```diff
@@ -32,123 +39,285 @@ def test_iterator(self):
...
+        env = self.env[0]
+        exp = self.source.env_step(0, env, actions[0])
+
+        self.assertTrue(isinstance(exp, Experience))
+
+    def test_source_next_single_env_single_step(self):
+        """Test that steps are executed correctly with one environment and 1 step"""
+
+        self.env = [gym.make("CartPole-v0") for _ in range(1)]
+        self.device = torch.device('cpu')
```
Comment (Borda) [inline tests/datamodules/test_experience_sources.py:102]: do we need to have explicitly CPU or let's take the best available?

File: tests/datamodules/test_experience_sources.py [hunk 4]
```diff
@@ -32,123 +39,285 @@ def test_iterator(self):
...
+            self.assertTrue(isinstance(exp, tuple))
+            self.assertTrue(len(exp) == self.source.n_steps)
+            break
+
+    def test_source_update_state(self):
+        """Test that after a step the state is updated"""
+
+        self.env = [gym.make("CartPole-v0") for _ in range(1)]
+        self.device = torch.device('cpu')
+        self.source = ExperienceSource(self.env, self.agent, n_steps=2)
```
Comment (Borda) [inline tests/datamodules/test_experience_sources.py:151]: you can move it to `SetUp` as it for most methods...

File: tests/datamodules/test_experience_sources.py [hunk 5]
```diff
@@ -32,123 +39,285 @@ def test_iterator(self):
...
+        n_envs = 2
+
+        self.env = [env1 for _ in range(2)]
+        self.device = torch.device('cpu')
+        self.source = ExperienceSource(self.env, self.agent, n_steps=3)
 
+        history = self.source.histories[0]
+        history.append(exp1)
+        history.append(exp2)
+        history.append(exp2)
```
Comment (Borda) [inline tests/datamodules/test_experience_sources.py:242]:
```suggestion
        history += [exp1, exp2, exp2]
```
really twice the same exp2?

File: tests/datamodules/test_experience_sources.py [hunk 6]
```diff
@@ -32,123 +39,285 @@ def test_iterator(self):
...
+
+        self.device = torch.device('cpu')
+        self.source = DiscountedExperienceSource(self.env1, self.agent, n_steps=self.n_steps)
+
+        self.source.histories[0].append(self.exp1)
+        self.source.histories[0].append(self.exp2)
+
+        for idx, exp in enumerate(self.source.stepper(self.device)):
+            self.assertTrue(isinstance(exp, Experience))
+            self.assertEqual(exp.reward, 2.9701)
```
Comment (Borda) [inline tests/datamodules/test_experience_sources.py:322]: mind add a comment on how the number comes up?

File: tests/models/test_rl/unit/test_agents.py [hunk 7]
```diff
@@ -18,43 +18,44 @@ def setUp(self) -> None:
...
         self.value_agent = ValueAgent(self.net, self.env.action_space.n)
 
     def test_value_agent(self):
 
         action = self.value_agent(self.state, self.device)
-        self.assertIsInstance(action, int)
+        self.assertIsInstance(action, list)
+        self.assertIsInstance(action[0], int)
 
     def test_value_agent_GET_ACTION(self):
```
Comment (Borda) [inline tests/models/test_rl/unit/test_agents.py:39]: pls no capital in function name...

File: tests/models/test_rl/unit/test_reinforce.py [hunk 8]
```diff
@@ -23,8 +21,7 @@ def setUp(self) -> None:
         self.n_actions = self.env.action_space.n
         self.net = MLP(self.obs_shape, self.n_actions)
         self.agent = Agent(self.net)
-        self.xp_stream = EpisodicExperienceStream(self.env, self.agent, Mock(), episodes=4)
-        self.rl_dataloader = DataLoader(self.xp_stream)
+        self.exp_source = DiscountedExperienceSource(self.env, self.agent)
 
         parent_parser = argparse.ArgumentParser(add_help=False)
```
Comment (Borda) [inline tests/models/test_rl/unit/test_reinforce.py:26]: this script test shall be in `tests/scripts/...` @ananyahjha93 ?

File: pl_bolts/datamodules/experience_source.py [hunk 9]
```diff
@@ -29,174 +31,247 @@ def __iter__(self) -> Iterable:
...
+    def __init__(self, env, agent) -> None:
         self.env = env
         self.agent = agent
-        self.state = self.env.reset()
-
-    def _reset(self) -> None:
-        """resets the env and state"""
-        self.state = self.env.reset()
-
-    def step(self, device: torch.device) -> Tuple[Experience, float, bool]:
```
[the remark this answers is not in the document]
Comment (djbyrne) [reply pl_bolts/datamodules/experience_source.py:53]: As in it doesn't need to be an ABC ?

File: pl_bolts/datamodules/experience_source.py [hunk 10]
```diff
@@ -29,174 +31,247 @@ def __iter__(self) -> Iterable:
...
-        self.gamma = gamma
+
+        if isinstance(env, (list, tuple)):
+            self.pool = env
+        else:
+            self.pool = [env]
+
         self.n_steps = n_steps
-        self.n_step_buffer = deque(maxlen=n_steps)
+        self.total_rewards = []
```
[the remark this answers is not in the document]
Comment (djbyrne) [reply pl_bolts/datamodules/experience_source.py:70]: As in should the total_rewards be a protected variable?

File: pl_bolts/datamodules/experience_source.py [hunk 11]
```diff
@@ -29,174 +31,247 @@ def __iter__(self) -> Iterable:
...
 
         Returns:
-            Experience
+            Tuple of Experiences
         """
-        exp = self.n_step(device)
+        while True:
+
+            # get actions for all envs
+            actions = self.env_actions(device)
```
[the remark this answers is not in the document]
Comment (djbyrne) [reply pl_bolts/datamodules/experience_source.py:93]: I still haven't worked on this with DP/DDP in mind. But my reasoning is that the device is set based on where the stepper is called. So it could be run across multiple devices in parallel. However in that case, the environments need to be split up. More work needs to be done on this

File: pl_bolts/datamodules/experience_source.py [hunk 12]
```diff
@@ -29,174 +31,247 @@ def __iter__(self) -> Iterable:
...
+                    yield tuple(history)
+
+                if exp.done:
+                    if 0 < len(history) < self.n_steps:
+                        yield tuple(history)
+
+                    # generate tail of history
+                    while len(history) > 2:
+                        history.popleft()
+                        yield tuple(history)
```
[the remark this answers is not in the document]
Comment (djbyrne) [reply pl_bolts/datamodules/experience_source.py:113]: Yeah, maybe not. I like that it allows us to retain the state of what environment is being sampled from and the history being built up. I suppose the alternative would be to maintain variables for what environment and what state was last sampled from. Any suggestions?

File: pl_bolts/datamodules/experience_source.py [hunk 13]
```diff
@@ -29,174 +31,247 @@ def __iter__(self) -> Iterable:
...
 
-        reward, next_state, done = self.get_transition_info()
-        first_experience = self.n_step_buffer[0]
-        multi_step_experience = Experience(
-            first_experience.state, first_experience.action, reward, done, next_state
-        )
+                    history.clear()
 
-        return multi_step_experience, exp.reward, exp.done
+            self.iter_idx += 1
```
[the remark this answers is not in the document]
Comment (djbyrne) [reply pl_bolts/datamodules/experience_source.py:123]: like `self.iter_idx_inf` ?

File: tests/models/test_rl/unit/test_agents.py (hunk 7 above)
Comment (djbyrne) [reply tests/models/test_rl/unit/test_agents.py:39]: No problem, i am used to having a main test name and then edge cases in caps. Ill change it

File: tests/datamodules/test_experience_sources.py (hunk 6 above)
Comment (djbyrne) [reply tests/datamodules/test_experience_sources.py:322]: will do

File: pl_bolts/datamodules/experience_source.py [hunk 14]
```diff
@@ -29,174 +31,247 @@ def __iter__(self) -> Iterable:
...
+        Returns:
+            list of total rewards for all completed episodes for each environment since last pop
+        """
+        rewards = self.total_rewards
 
-    def __init__(self, env, agent, device: torch.device, episodes: int = 1):
-        super().__init__(env, agent)
-        self.episodes = episodes
-        self.device = device
+        if rewards:
```
[the remark this answers is not in the document]
Comment (djbyrne) [reply pl_bolts/datamodules/experience_source.py:196]: The total_rewards is a list that contains the latest rewards at the end of an episode. This gets called at the end of each step in our model. If the model calls pop_rewards_step, if an episode has finished, total_rewards will contain a value or if multiple episodes have finished since we last check, there will be multiple values. These get store in reward, then  we clear the total_rewards. We only want to empty self.total_rewards once we have taken the latest values. Does that make sense?

File: pl_bolts/models/rl/dqn_model.py [hunk 15]
```diff
@@ -149,75 +157,80 @@ def build_networks(self) -> None:
...
         return output
 
+    def train_batch(self) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+        """
+        Contains the logic for generating a new batch of data to be passed to the DataLoader
+        Returns:
+            yields a Experience tuple containing the state, action, reward, done and next_state.
+        """
+
+        for step_idx, exp in enumerate(self.source.stepper(self.device)):
```
[the remark this answers is not in the document]
Comment (djbyrne) [reply pl_bolts/models/rl/dqn_model.py:178]: The train_batch would be infinite yes because the data is constantly being generated

File: pl_bolts/models/rl/per_dqn_model.py [hunk 16]
```diff
@@ -73,68 +109,55 @@ def training_step(self, batch, _) -> OrderedDict:
...
 
-        # Soft update of target network
+        # update of target network
         if self.global_step % self.sync_rate == 0:
             self.target_net.load_state_dict(self.net.state_dict())
 
         log = {
-            "total_reward": torch.tensor(self.total_reward).to(self.device),
-            "avg_reward": torch.tensor(self.avg_reward),
+            "total_reward": self.total_rewards[-1],
```
[the remark this answers is not in the document]
Comment (djbyrne) [reply pl_bolts/models/rl/per_dqn_model.py:128]: Ah, reward_sum should be removed, it is not being used anymore. The reason is that reward_sum is not accurate, if an episode ended when the first datum of a 32 size mini batch is being pulled, the reward_sum would get reset. By the time it gets to the training step logging, reward_sum would be the total reward of a new episode. I will update this

File: pl_bolts/models/rl/reinforce_model.py [hunk 17]
```diff
@@ -115,122 +102,69 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
-            List of qvals for each episodes
+            list of discounted rewards
         """
+        assert isinstance(rewards[0], float)
+
         res = []
         sum_r = 0.0
-        for reward in reversed(rewards):
+        for r in reversed(rewards):
             sum_r *= self.gamma
```
[the remark this answers is not in the document]
Comment (djbyrne) [reply pl_bolts/models/rl/reinforce_model.py:133]: I tried a few variations of single line eq here, but list comprehension does not work well as the sum_r must be accumulated. Then only other option would be to use a lambda function, but in my opinion this makes its less comprehensible

File: pl_bolts/models/rl/reinforce_model.py [hunk 18]
```diff
@@ -115,122 +102,69 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
-        # get qvals
-        batch_qvals = []
-        for reward in batch_rewards:
-            batch_qvals.append(self.calc_qvals(reward))
-
-        # flatten the batched outputs
-        batch_actions, batch_qvals, batch_rewards, batch_states = self.flatten_batch(
-            batch_actions, batch_qvals, batch_rewards, batch_states
-        )
+        for step_idx, exp in enumerate(self.exp_source.stepper(self.device)):
```
[the remark this answers is not in the document]
Comment (djbyrne) [reply pl_bolts/models/rl/reinforce_model.py:147]: no its not, ill change it to _

File: tests/datamodules/test_experience_sources.py (hunk 2 above)
Comment (djbyrne) [reply tests/datamodules/test_experience_sources.py:82]: im not sure what you mean

File: tests/datamodules/test_experience_sources.py (hunk 3 above)
Comment (djbyrne) [reply tests/datamodules/test_experience_sources.py]: No, it should take the best available, ill update it

File: pl_bolts/datamodules/experience_source.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Borda) [reply pl_bolts/datamodules/experience_source.py:70]: just an idea, I would make it `_total_rewards`

File: pl_bolts/datamodules/experience_source.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Borda) [reply pl_bolts/datamodules/experience_source.py:113]: then rather have some logic about determining the value, but have `yield` only once

File: pl_bolts/datamodules/experience_source.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (Borda) [reply pl_bolts/datamodules/experience_source.py:123]: I was thinking about rename the function instead...

File: tests/datamodules/test_experience_sources.py (hunk 2 above)
Comment (Borda) [reply tests/datamodules/test_experience_sources.py:82]: that in the test init you already have `self.net = Mock()` so here you can use `self.net` instead of `single_env`

File: pl_bolts/datamodules/experience_source.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (djbyrne) [reply pl_bolts/datamodules/experience_source.py:113]: ok, ill get on it

File: pl_bolts/models/rl/reinforce_model.py [hunk 19]
```diff
@@ -115,122 +102,69 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
-        batch_rewards: List[List[Tensor]],
-        batch_states: List[List[Tuple[Tensor, Tensor]]],
-    ) -> Tuple[Tensor, Tensor, Tensor, Tensor]:
-        """
-        Takes in the outputs of the processed batch and flattens the several episodes into a single tensor for each
-        batched output
+            if exp.done:
+                self.batch_qvals.extend(self.calc_qvals(self.cur_rewards))
+                self.cur_rewards.clear()
+                self.batch_episodes += 1
```
[the remark this answers is not in the document]
Comment (djbyrne) [reply pl_bolts/models/rl/reinforce_model.py:158]: Yes, the environment will be reset in the experience_source. This just needs to update the episode based variables

**Comment by Borda** [comment]
lets merge this one, it is quite large...
@williamFalcon mind check it?

File: pl_bolts/datamodules/experience_source.py [hunk 20]
```diff
@@ -29,174 +31,265 @@ def __iter__(self) -> Iterable:
...
+        the done state, the history will be incrementally appended to the queue, removing the tail of the history
+        each time.
+        Args:
+            env_idx: index of the environment
+            exp: the current experience
+            history: history of experience steps for this environment
+        """
+        # If there is a full history of step, append history to queue
+        if len(history) == self.n_steps:
+            self.exp_history_queue.append(tuple(history))
```
Comment (Borda) [inline pl_bolts/datamodules/experience_source.py:124]: does has to e tuple?

File: pl_bolts/datamodules/experience_source.py [hunk 21]
```diff
@@ -29,174 +31,265 @@ def __iter__(self) -> Iterable:
...
+            if len(history) > 1:
+                self.update_env_stats(env_idx)
+
+                history.popleft()
+                self.exp_history_queue.append(tuple(history))
+
+            # Clear that last tail in the history once all others have been added to the queue
+            history.clear()
+
+    def init_envs(self) -> None:
```
Comment (Borda) [inline pl_bolts/datamodules/experience_source.py:146]:
```suggestion
    def init_environemnts(self) -> None:
```

File: pl_bolts/datamodules/__init__.py [hunk 22]
```diff
@@ -6,6 +6,6 @@
 from pl_bolts.datamodules.stl10_datamodule import STL10DataModule
 from pl_bolts.datamodules.fashion_mnist_datamodule import FashionMNISTDataModule
 from pl_bolts.datamodules.dummy_dataset import DummyDataset
-from pl_bolts.datamodules.experience_source import (ExperienceSourceDataset, ExperienceSource,
-                                                    NStepExperienceSource, EpisodicExperienceStream)
 from pl_bolts.datamodules.async_dataloader import AsynchronousLoader
+from pl_bolts.datamodules.experience_source import (ExperienceSourceDataset, ExperienceSource,
+                                                    DiscountedExperienceSource)
```
Comment (Borda) [inline pl_bolts/datamodules/__init__.py:11]:
```suggestion
from pl_bolts.datamodules.experience_source import (
    ExperienceSourceDataset, ExperienceSource, DiscountedExperienceSource)
```

File: pl_bolts/models/rl/common/agents.py [hunk 23]
```diff
@@ -83,11 +89,11 @@ def get_action(self, state: torch.Tensor, device: torch.device):
...
         """
         if not isinstance(state, torch.Tensor):
-            state = torch.tensor([state], device=device)
+            state = torch.tensor(state, device=device)
 
         q_values = self.net(state)
-        _, action = torch.max(q_values, dim=1)
-        return int(action.detach().item())
+        _, actions = torch.max(q_values, dim=1)
+        return actions.detach().cpu().numpy()
```
Comment (Borda) [inline pl_bolts/models/rl/common/agents.py:96]: can we keep it on the device for higher speed?

File: pl_bolts/models/rl/common/experience.py [hunk 24]
```diff
@@ -29,8 +29,10 @@ class RLDataset(IterableDataset):
     """
 
     def __init__(self, buffer: Buffer, sample_size: int = 1) -> None:
-        warnings.warn("Deprecated, these functions have been moved to pl_bolts.datamodules.experience_source.py",
-                      DeprecationWarning)
+        warnings.warn(
```
Comment (Borda) [inline pl_bolts/models/rl/common/experience.py:32]: add info when it will be removed

File: pl_bolts/models/rl/dqn_model.py [hunk 25]
```diff
@@ -242,29 +270,45 @@ def configure_optimizers(self) -> List[Optimizer]:
...
-        return DataLoader(dataset=self.dataset, batch_size=self.batch_size)
+        return self._dataloader()
 
     def test_dataloader(self) -> DataLoader:
         """Get test loader"""
         return DataLoader(dataset=self.dataset, batch_size=self.batch_size)
 
     @staticmethod
-    def add_model_specific_args(arg_parser: argparse.ArgumentParser) -> argparse.ArgumentParser:
+    def make_env(env_name: str, seed: int) -> gym.Env:
```
Comment (Borda) [inline pl_bolts/models/rl/dqn_model.py:290]:
```suggestion
    def make_ environment(env_name: str, seed: int) -> gym.Env:
```
is create/prepare better word?

File: pl_bolts/models/rl/reinforce_model.py [hunk 26]
```diff
@@ -115,122 +116,80 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        sum_r = 0.0
 
-        # stack steps into single tensor and remove extra dimension
-        batch_qvals = torch.stack(batch_qvals).squeeze()
-        batch_states = torch.stack(batch_states).squeeze()
-        batch_actions = torch.stack(batch_actions).squeeze()
-        batch_rewards = torch.stack(batch_rewards).squeeze()
+        for r in reversed(rewards):
+            sum_r *= self.gamma
+            sum_r += r
```
Comment (Borda) [inline pl_bolts/models/rl/reinforce_model.py:135]:
```suggestion
            sum_r = (sum_r * self.gamma) + r
```

File: pl_bolts/models/rl/reinforce_model.py [hunk 27]
```diff
@@ -115,122 +116,80 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
-        Calculates the mse loss using a batch of states, actions and Q values from several episodes. These have all
-        been flattend into a single tensor.
-
-        Args:
-            batch_qvals: current mini batch of q values
-            batch_actions: current batch of actions
-            batch_states: current batch of states
+        Contains the logic for generating a new batch of data to be passed to the DataLoader
 
         Returns:
```
Comment (Borda) [inline pl_bolts/models/rl/reinforce_model.py:146]: you should use the key work
```suggestion
        Yield:
```
https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html#docstring-sections

File: tests/datamodules/test_experience_sources.py [hunk 28]
```diff
@@ -5,22 +5,33 @@
...
+    DiscountedExperienceSource,
+)
 from pl_bolts.models.rl.common.agents import Agent
 
 
 class DummyAgent(Agent):
-    def __call__(self, states, agent_states):
-        return 0
+    def __call__(self, states, device):
+        return [0 for s in states]
```
Comment (Borda) [inline tests/datamodules/test_experience_sources.py:20]:
```suggestion
        return [0] * len(states)
```

File: pl_bolts/models/rl/dqn_model.py (hunk 25 above)
Comment (djbyrne) [reply pl_bolts/models/rl/dqn_model.py:290]: i think make_environment is fine, but Im ok with changing it

File: pl_bolts/datamodules/experience_source.py (hunk 20 above)
Comment (djbyrne) [reply pl_bolts/datamodules/experience_source.py:124]: yeah, the history is sliced later on and you can't slice a deque

File: pl_bolts/models/rl/common/experience.py (hunk 24 above)
Comment (djbyrne) [reply pl_bolts/models/rl/common/experience.py:32]: when should it be removed?

File: pl_bolts/models/rl/common/agents.py (hunk 23 above)
Comment (djbyrne) [reply pl_bolts/models/rl/common/agents.py:96]: the action cant be a tensor as it is passed to the gym environment

File: pl_bolts/models/rl/common/experience.py (hunk 24 above)
Comment (Borda) [reply pl_bolts/models/rl/common/experience.py:32]: as there is not any release yet I would drop this message completely

Comment (djbyrne) [reply pl_bolts/models/rl/common/experience.py]: I have removed this as It shouldn't be used going forward and could cause compatibility issues if left in

File: pl_bolts/datamodules/experience_source.py (hunk 21 above)
Comment (djbyrne) [reply pl_bolts/datamodules/experience_source.py]: done

**Comment by Borda** [comment]
LGTM, @williamFalcon mind do final check?
